File: frontend/src/api/satellite.ts
import { api } from './client'
import type { Event, SatelliteArea, SatelliteMonitorResult, SatelliteScene, SatelliteStats } from '../types'

export const satelliteApi = {
  // ── Areas ────────────────────────────────────────────────────────────────
  listAreas: () => api.get<SatelliteArea[]>('/satellite/areas'),

  createArea: (data: {
    name: string
    description?: string | null
    geojson: Record<string, unknown>
    enabled?: boolean
  }) => api.post<SatelliteArea>('/satellite/areas', data),

  updateArea: (id: number, data: {
    name?: string
    description?: string | null
    geojson?: Record<string, unknown>
    enabled?: boolean
  }) => api.patch<SatelliteArea>(`/satellite/areas/${id}`, data),

  deleteArea: (id: number) => api.delete(`/satellite/areas/${id}`),

  // ── Scenes ───────────────────────────────────────────────────────────────
  listScenes: (params?: {
    area_id?: number
    mission?: string
    max_cloud_cover?: number
    date_from?: string
    date_to?: string
    limit?: number
    offset?: number
  }) => {
    const q = new URLSearchParams()
    if (params?.area_id != null) q.set('area_id', String(params.area_id))
    if (params?.mission) q.set('mission', params.mission)
    if (params?.max_cloud_cover != null) q.set('max_cloud_cover', String(params.max_cloud_cover))
    if (params?.date_from) q.set('date_from', params.date_from)
    if (params?.date_to) q.set('date_to', params.date_to)
    if (params?.limit != null) q.set('limit', String(params.limit))
    if (params?.offset != null) q.set('offset', String(params.offset))
    return api.get<SatelliteScene[]>(`/satellite/scenes${q.size ? `?${q}` : ''}`)
  },

  createScene: (data: Partial<SatelliteScene> & {
    external_id: string
    bbox: number[]
    footprint: Record<string, unknown>
  }) => api.post<SatelliteScene>('/satellite/scenes', data),

  deleteScene: (id: number) => api.delete(`/satellite/scenes/${id}`),

  importStac: (data: {
    item?: Record<string, unknown>
    items?: Record<string, unknown>[]
    feature_collection?: Record<string, unknown>
    area_id?: number | null
    skip_existing?: boolean
  }) => api.post<SatelliteScene[]>('/satellite/scenes/import-stac', data),

  /** Thumbnail proxy URL (call directly as <img src> or fetch). */
  thumbnailUrl: (sceneId: number): string =>
    `/api/satellite/scenes/${sceneId}/thumbnail`,

  // ── STAC search ──────────────────────────────────────────────────────────
  searchStac: (payload: {
    url: string
    auth_token?: string
    [key: string]: unknown
  }) => api.post<{ type: string; features: Record<string, unknown>[] }>('/satellite/search-stac', payload),

  searchAndImport: (data: {
    url: string
    collections?: string[]
    bbox?: number[]
    date_from?: string
    date_to?: string
    max_cloud_cover?: number
    limit?: number
    area_id?: number | null
    auth_token?: string
    skip_existing?: boolean
  }) => api.post<SatelliteScene[]>('/satellite/search-and-import', data),

  // ── Monitor ──────────────────────────────────────────────────────────────
  runMonitor: (data: {
    area_id?: number | null
    max_cloud_cover?: number | null
    create_events?: boolean
  }) => api.post<SatelliteMonitorResult>('/satellite/monitor/run', data),

  listEvents: (limit = 200) =>
    api.get<Event[]>(`/satellite/events?limit=${limit}`),

  stats: () => api.get<SatelliteStats>('/satellite/stats'),
}

File: backend/app/schemas/satellite.py
from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class SatelliteAreaCreate(BaseModel):
    name: str
    description: str | None = None
    geojson: dict[str, Any]
    enabled: bool = True


class SatelliteAreaUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    geojson: dict[str, Any] | None = None
    enabled: bool | None = None


class SatelliteAreaRead(BaseModel):
    id: int
    name: str
    description: str | None = None
    geojson: dict[str, Any]
    bbox: list[float]
    enabled: bool
    created_at: datetime
    scene_count: int = 0

    model_config = {"from_attributes": True}


class SatelliteSceneCreate(BaseModel):
    external_id: str
    provider: str = "manual"
    mission: str | None = None
    product_type: str | None = None
    acquired_at: datetime | None = None
    cloud_cover: float | None = Field(default=None, ge=0, le=100)
    bbox: list[float]
    footprint: dict[str, Any]
    assets: dict[str, Any] = Field(default_factory=dict)
    metadata: dict[str, Any] | None = None
    local_path: str | None = None
    thumbnail_url: str | None = None
    source_url: str | None = None
    area_id: int | None = None
    status: str = "available"


class SatelliteSceneUpdate(BaseModel):
    external_id: str | None = None
    provider: str | None = None
    mission: str | None = None
    product_type: str | None = None
    acquired_at: datetime | None = None
    cloud_cover: float | None = Field(default=None, ge=0, le=100)
    bbox: list[float] | None = None
    footprint: dict[str, Any] | None = None
    assets: dict[str, Any] | None = None
    metadata: dict[str, Any] | None = None
    local_path: str | None = None
    thumbnail_url: str | None = None
    source_url: str | None = None
    area_id: int | None = None
    status: str | None = None


class SatelliteSceneRead(BaseModel):
    id: int
    external_id: str
    provider: str
    mission: str | None = None
    product_type: str | None = None
    acquired_at: datetime | None = None
    cloud_cover: float | None = None
    bbox: list[float]
    footprint: dict[str, Any]
    assets: dict[str, Any]
    metadata: dict[str, Any] | None = Field(default=None, validation_alias="metadata_")
    local_path: str | None = None
    thumbnail_url: str | None = None
    source_url: str | None = None
    area_id: int | None = None
    status: str
    created_at: datetime

    model_config = {"from_attributes": True, "populate_by_name": True}


class SatelliteStacImportRequest(BaseModel):
    item: dict[str, Any] | None = None
    items: list[dict[str, Any]] | None = None
    feature_collection: dict[str, Any] | None = None
    area_id: int | None = None
    skip_existing: bool = True


class StacSearchAndImportRequest(BaseModel):
    """Search a STAC endpoint and import matching scenes directly."""
    url: str
    collections: list[str] = Field(default_factory=list)
    bbox: list[float] | None = None
    date_from: str | None = None
    date_to: str | None = None
    max_cloud_cover: float | None = Field(default=None, ge=0, le=100)
    limit: int = Field(default=50, ge=1, le=200)
    area_id: int | None = None
    auth_token: str | None = None
    skip_existing: bool = True


class SatelliteMonitorRequest(BaseModel):
    area_id: int | None = None
    max_cloud_cover: float | None = Field(default=30, ge=0, le=100)
    create_events: bool = True


class SatelliteMonitorResult(BaseModel):
    matched_scenes: int
    created_events: int


class SatelliteStats(BaseModel):
    areas: int
    scenes: int
    events: int
    by_mission: dict[str, int]
